### DISCUSSION TITLE
Type hinting with BaseModel hierarchies

### DISCUSSION THREAD

**Discussion by LucaBonfiglioli** [pydantic/pydantic#4208] [Question] [answered]
Hello, just a quick question about class hierarchies and type hinting that has been troubling me. 

Suppose I have the following pretty simple class hierarchy:
```python
from abc import ABC, abstractmethod

from pydantic import BaseModel
from typing import Literal, Union


class Base(BaseModel, ABC):
    a: int = 10
    b: float = 0.5

    @abstractmethod
    def foo(self):
        pass


class Derived1(Base):
    type_: Literal["derived1"] = "derived1"

    def foo(self):
        print("AAAAA")


class Derived2(Base):
    type_: Literal["derived2"] = "derived2"

    def foo(self):
        print("BBBBB")
```

Then, I want to use a model of type `Base` inside another pydantic model named `Client`. This is a pretty common structure that occurs pretty much everywhere. 

```python
class Client(BaseModel):
    base: Base
    # base: Union[Derived1, Derived2] 

    def foo(self):
        self.base.foo()
```

When parsing a dictionary, it always raises an error. Code and error below:

```python
if __name__ == '__main__': 
    data = {
        "base": {
            "a": 10,
            "b": 0.5,
            "type_": "derived2",
        }
    }
    
    client = Client.parse_obj(data)
    client.foo()
```  

```
Traceback (most recent call last):
  File "pydantic_example.py", line 47, in <module>
    client = Client.parse_obj(data)
  File "pydantic\main.py", line 521, in pydantic.main.BaseModel.parse_obj
  File "pydantic\main.py", line 341, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Client
base
  Can't instantiate abstract class Base with abstract methods foo (type=type_error)
```

Apparently it tries to instantiate the model `Base`, which is an abstract class, so of course it raises an error! Why though? Wouldn't be much simpler to look for a suitable subtype (in this case `Derived2`) when trying to instantiate an ABC?

The example works if I change the type hinting from `Base` to `Union[Derived1, Derived2]`, but doing so removes all the benefits of proper type hinting, it is not transparent, it is unnecessarily verbose and the signature must be updated whenever a new subtype of `Base` is added. Is there another way to do this, or am I trying to force something that should not be done?

**Comment by alcrene** [comment, marked as answer]
The problem here is, for this to work as you've suggested, `Base` would need to know its subclasses, and that's just not how subclassing works: child classes know their parents, but parents don't keep track of their children. This is still true for abstract classes.
So to do what you want, you would need a type (say `AbstractBaseModel`) which updates an internal registry whenever a subclass is created, and then searches this registry during validation for the right type to use. This is rather tricky to get right (especially if hierarchies can have multiple levels), which I can attest to because I had a similar problem and it took me five implementations before obtaining [something satisfactory](https://github.com/alcrene/scityping#an-extensible-serialization-hierarchy).

Baring a full blown solution like the one I linked to, if you are mostly worried about readability, it might suffice to define the union type (I would put this in the same module as the definitions of `Derivedx`):
```python
# mytypes1.py
...
AnyBase = Union[Derived1, Derived2]
```
Since multiple `Union`s get collapsed together, another module can extend the definition:
```python
# mytypes2.py
import mytypes1

class Derived3(mytypes1.Base):
  ...

AnyBase = Union[mytypes1.AnyBase, Derived3]
```
To have a consistent source for `AnyBase`, you could even then do `mytypes1.AnyBase = AnyBase` – whether monkey patching `mytypes1` like that is acceptable will depend on your use case.

As a minor comment regarding your example: by default pydantic will silently ignore extra keywords, which is why the validation on `Base` succeeds despite the `type_` keyword. You can [change this](https://pydantic-docs.helpmanual.io/usage/model_config/), but in this case also defining `type_` on `Base` with a different value probably makes more sense.
